Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,176 @@
+* COM-style ABI in `coreclr`  
+`coreclr` exposes COM-style ABI to host the .NET Core runtime and perform a wide range of operations on it. See this [header file](https://github.com/dotnet/coreclr/blob/master/src/pal/prebuilt/inc/mscoree.h) for more details.
+Similarly to the C-style ABI the COM-style ABI also requires the native app to locate the runtime and to fully specify all startup parameters.
+There's no inherent interoperability between these APIs and the .NET Core SDK.
+* `hostfxr` and `hostpolicy` APIs
+The hosting layer of .NET Core already exposes some functionality as C-style ABI on either the `hostfxr` or `hostpolicy` libraries. These can execute application, determine available SDKs, determine native dependency locations, resolve component dependencies and so on.
+Unlike the above `coreclr` based APIs these don't require the caller to fully specify all startup parameters, instead these APIs understand artifacts produced by .NET Core SDK making it much easier to consume SDK produced apps/libraries.
+The native app is still required to locate the `hostfxr` or `hostpolicy` libraries. These APIs are also designed for specific narrow scenarios, any usage outside of these bounds is typically not possible.
+
+
+## Scope
+This document focuses on easy-to-use hosting which cooperates with the .NET Core SDK and consumes the artifacts produced by building the managed app/libraries directly. It completely ignores the COM-style ABI as it's hard to use from some programming languages.
+
+As such the document explicitly excludes any hosting based on directly loading `coreclr`. The document focuses on using the existing .NET Core hosting layer in new ways. For details on the .NET Core hosting components see [this document](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md).
+
+
+## High-level proposal
+In .NET Core 3.0 the hosting layer (see [here](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md)) ships with several hosts. These are binaries which act as the entry points to the .NET Core hosting/runtime:
+* The "muxer" (`dotnet.exe`)
+* The `apphost` (`.exe` which is part of the app)
+* The `comhost` (`.dll` which is part of the app and acts as COM server)
+* The `ijwhost` (`.dll` consumed via `.lib` used by IJW assemblies)
+
+Every one of these hosts serve different scenario and expose different APIs. The one thing they have in common is that their main purpose is to find the right `hostfxr`, load it and call into it to execute the desired scenario. For the most part all these hosts are basically just wrappers around functionality provided by `hostfxr`.
+
+The proposal is to add a new host library `nethost` which can be used by native apps to easily host managed components and to easily locate `hostfxr` for more advanced scenarios.
```

Comment:
What's the gesture to get `nethost` into customers app published output?

Reply:
I don't have the details. The problem is that this goes outside or our ecosystem. Basically this is about C++ projects or even languages which MS has nothing to do with (Python, Java, ...).
For now I'm thinking just a NuGet package if it can handle the per-platform nicely. But ultimetly we'll either need to invest in language specific packaging, or just expose pure zips or something like that.

Reply:
We will be shipping the `.zip`. We're investigating adding NuGet as well, but there are some issues with this, so no commitment on that.